### PR TITLE
(#3603) add more application functions

### PR DESCRIPTION
**Pull request by revdan**
I've added in the functions mentioned in #3603 plus `:application.get_all_key/1`, but while I'm looking at it are there any others you'd like in here? (as you specified 'at least' those three)

Also I'd be very happy with any feedback on the typespecs/docs and the tests themselves, as I wasn't sure of the best/safest things to test. 

Thanks! :smile:

### CODE REVIEW

**Comment by lexmag** [comment]
Hey @revdan, there is already a PR (#3604) that adds `get_all_key/1` and `get_key/3` but not `started_applications/0` and `loaded_applications/0` thought.

**Comment by revdan** [comment]
Thanks @lexmag, I didn't spot that :)

Will remove those parts and re-raise.

**Comment by revdan** [comment]
Well that was dumb. I'd still like feedback, but should I also squash this back into one commit?

File: lib/elixir/lib/application.ex [hunk 1]
```diff
@@ -334,6 +334,22 @@ defmodule Application do
   end
 
   @doc """
+  Returns a list with information about the applications which are currently running.
+  """
+  @spec started_applications :: [tuple]
+  def started_applications do
```
Comment (lexmag) [inline lib/elixir/lib/application.ex]: I think it should be `started_applications/1` with default timeout value and call `:application.which_applications/1`.

File: lib/elixir/test/elixir/application_test.exs [hunk 2]
```diff
@@ -19,6 +19,15 @@ defmodule ApplicationTest do
 
     assert Application.delete_env(:elixir, :unknown) == :ok
     assert Application.get_env(:elixir, :unknown, :default) == :default
+
```
Comment (lexmag) [inline lib/elixir/test/elixir/application_test.exs]: Needless extra line.

File: lib/elixir/lib/application.ex (hunk 1 above)
Comment (revdan) [inline lib/elixir/lib/application.ex]: #3603 specifies `:application.which_applications/0` but I'm happy to add in `/1` as well :)

File: lib/elixir/test/elixir/application_test.exs (hunk 2 above)
Comment (revdan) [inline lib/elixir/test/elixir/application_test.exs]: Fair point, my bad.

File: lib/elixir/lib/application.ex (hunk 1 above)
Comment (lexmag) [inline lib/elixir/lib/application.ex]: Having `/1` with default value you'll get `/0` automatically. :)

Comment (revdan) [inline lib/elixir/lib/application.ex]: Now that's using your noggin :+1:

File: lib/elixir/test/elixir/application_test.exs [hunk 3]
```diff
@@ -19,6 +19,15 @@ defmodule ApplicationTest do
...
+  end
+
+  test "application information" do
+    assert is_list(Application.started_applications)
```
Comment (josevalim) [inline lib/elixir/test/elixir/application_test.exs]: You should check it is a list and then use `List.keyfind/3` to get Elixir out of it and assert on the returned values.

File: lib/elixir/test/elixir/application_test.exs [hunk 4]
```diff
@@ -19,6 +19,15 @@ defmodule ApplicationTest do
...
+    assert is_tuple(List.first(Application.started_applications))
+
+    assert is_list(Application.loaded_applications)
+    assert is_tuple(List.first(Application.loaded_applications))
```
Comment (josevalim) [inline lib/elixir/test/elixir/application_test.exs]: Same here.

**Comment by lexmag** [comment]
> should I also squash this back into one commit?
Yes please. :smile:

File: lib/elixir/test/elixir/application_test.exs (hunk 4 above)
Comment (revdan) [inline lib/elixir/test/elixir/application_test.exs]: Thanks, doing both of those now!

**Comment by revdan** [comment]
Thanks so much for the feedback, @lexmag and @josevalim! Hopefully this is up to scratch now :)

File: lib/elixir/test/elixir/application_test.exs [hunk 5]
```diff
@@ -21,6 +21,20 @@ defmodule ApplicationTest do
     assert Application.get_env(:elixir, :unknown, :default) == :default
   end
 
+  test "application information" do
+    started = Application.started_applications
+    assert started |> is_list
```
Comment (lexmag) [inline lib/elixir/test/elixir/application_test.exs]: Just a small nitpicking: this pipe call looks excessive.

Comment (revdan) [inline lib/elixir/test/elixir/application_test.exs]: But I just love pipes :)

If that's not the code style here then happy to change though.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by revdan** [comment]
Thanks :grin::grin::grin:

File: lib/elixir/test/elixir/application_test.exs (hunk 5 above)
Comment (lexmag) [inline lib/elixir/test/elixir/application_test.exs]: Everyone loves pipes, [even too much](https://twitter.com/lexmag/status/618432396816457728). :smile:
It just feels excessive to use it for a single function call IMO.

Comment (revdan) [inline lib/elixir/test/elixir/application_test.exs]: Hah! The programming equivalent of dad saying he'll turn the car around if you don't shut up in the back :)

**Pull request merged**